### DISCUSSION TITLE
[Poetry & gitlab registries] can lock but can't install package from gitlab private registry

### DISCUSSION THREAD

**Discussion by ClementPinard** [python-poetry/poetry#7827] [Q&A]
- **Poetry version**: 1.4.2
- **Python version**: 3.10
- **OS version and name**: Ubuntu 22.04

Hello,

My team recently switched to a Gitlab private pypi registry. See https://docs.gitlab.com/ee/user/packages/pypi_repository/

From my understanding of how it works, we just need to change the source, use the token and we are good to go :

```bash
poetry source add my_pypi https://gitlab.com/api/v4/projects/[REDACTED]/packages/pypi/
poetry config pypi-token.my_pypi <token>
```

where `<token>` is the personal token with access to `read_api`

1st problem : poetry does not seem to find any package. That's okay, we can just use the `simple` API:

```bash
poetry source add my_pypi https://gitlab.com/api/v4/projects/[REDACTED]/packages/pypi/simple/
```

This works, but only for locking ! After that, the download is blocked :

```
poetry add my_lib

Using version ^1.1.12 for my_lib

Updating dependencies
Resolving dependencies... (0.5s)

Writing lock file

Package operations: 1 install, 0 updates, 0 removals

  • Installing my_lib (1.1.12): Failed

  HTTPError

  401 Client Error: Unauthorized for url: https://gitlab.com/api/v4/projects/[REDACTED]/packages/pypi/files/[REDACTED]/my_lib-1.1.12-py3-none-any.whl#sha256=[REDACTED]

  at ~/.local/share/pypoetry/venv/lib/python3.10/site-packages/requests/models.py:1021 in raise_for_status
      1017│                 f"{self.status_code} Server Error: {reason} for url: {self.url}"
      1018│             )
      1019│ 
      1020│         if http_error_msg:
    → 1021│             raise HTTPError(http_error_msg, response=self)
      1022│ 
      1023│     def close(self):
      1024│         """Releases the connection back to the pool. Once this method has been
      1025│         called the underlying ``raw`` object must not be accessed again.

```

From my investigation, it looks like that poetry can't download the file because it does not associate `my_pypi`'s credentials with the url. And it does not associated the credentials of source `my_pypi` because the url almost match the source url, but not exactly: it does not have the `/simple/` in it.

To verify that, doing the following commands makes it work :

```bash
poetry source add my_pypi2 https://gitlab.com/api/v4/projects/[REDACTED]/packages/pypi/
poetry config pypi-token.my_pypi2 <token>
```

(notice that pypi2 url does not have the trailing `/simple/`)

What do you think is happening here ? The workaround I proposed is not very pretty.

Is it gitlab that should have a better implementation of their simple API ? Is it normal download links don't have the `/simple/` in it ?
Should poetry try credentials from url but also url + `/simple/`  ?
Should I force another API than simple ? In gitlab's docs, for pip, it looks like they advice to add the registry url in the `~/.pypirc` without the trailing `/simple` https://docs.gitlab.com/ee/user/packages/pypi_repository/#authenticate-with-a-personal-access-token so maybe the JSON  API is inplemented somehow ?

Here is a similar discussion : https://github.com/orgs/python-poetry/discussions/7083 , the linked issue did help me uderstand my better, but it's not exactly related. Besides, the fix proposed is already present in last poetry version.

Thanks in advance,

Clément

**Comment by dimbleby** [comment]
`read_api` sounds like it probably doesn't give permission to download files?

https://github.com/python-poetry/poetry/blob/bc6e84e3fbcf071ddc61eb60f12b9fba83ec29e9/src/poetry/utils/authenticator.py#L419 looks like the code to investigate if you want to understand why the configuration is or isn't matching for a given url

**Comment by ClementPinard** [comment]
the token works with pip and with the two sources solution. So I don't think it's the problem.

Okay, the solution is apparently fixed (for now!)

After some awfully tedious investigation, it turns out that I had an empty "my_pypi" mention in my `~/.config/pypoetry/config.toml` file :

```toml
[repositories.my_pypi]
```

It's probably a legacy from before when I had it in my config rather than my `pyproject.toml` file. This added in my config an empty source with a path set to `None`. Since it does not have any url, it didn't match with anything and made the whole credential retrieval useless.

Would have been great to see it in when I call `poetry config --list`  but it didn't, I had to look directly in the toml fil to see it.

Anyway, it seems to be solved now. The weird part now is how I was able to still lock the file. The version chooser probably don't use the authenticator the same way I guess.

Thanks a lot for pointing me to this function, helped a ton, as I had to search why the `self.configured_repositories`  became empty after choosing the version to install.
